src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts

```diff
@@ -600,6 +600,9 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 				});
 			}
 		}).finally(() => {
+			const commonKey = task.getCommonTaskId();
+			liveDependencies.delete(commonKey);
```

Comment:
This isn't needed. The stack of live dependencies is maintained by cloning the set. I don't think it breaks anything, but it unnecessarily extends the lifetime of the set, preventing it from being garbage collected earlier.

Reply:
Run `start demo`. Kill the `watch` task. Run `start demo`. Without this change, and with the changes from this PR, it doesn't restart `watch`

```
{
  "version": "2.0.0",
  "presentation": {
    "echo": false,
    "reveal": "always",
    "focus": false,
    "panel": "dedicated",
    "showReuseMessage": true
  },
  "tasks": [
    {
      "type": "npm",
      "script": "test",
      "group":{
        "kind": "test",
        "isDefault": true
      },
      "problemMatcher": []
    },
    {
      "label": "watch",
      "type": "npm",
      "script": "watch",
      "group": "build",
      "isBackground": true,
      "problemMatcher": "$tsc-watch",
      "presentation": {
        "group": "vscode"
      }
    },
    {
      "label": "start",
      "type": "npm",
      "script": "start",
      "group": "build",
      "isBackground": true,
      "problemMatcher": [],
      "presentation": {
        "group": "vscode"
      }
    },
    {
      "label": "Start demo",
      "dependsOn": ["start", "watch"],
      "group": {
        "kind": "build",
        "isDefault": true
      },
      "isBackground": true,
      "problemMatcher": {
        "owner": "typescript",
        "fileLocation": [
          "relative",
          "${workspaceFolder}"
        ],
        "pattern": [
          {
            "regexp": "^([^\\\\s].*)\\\\((\\\\d+,\\\\d+)\\\\):\\\\s*(.*)$",
            "file": 1,
            "location": 2,
            "message": 3
          }
        ],
        "background": {
          "beginsPattern": "assets by",
          "endsPattern": "webpack \\d+\\.\\d+\\.\\d+ compiled successfully"
        }
      },
      "presentation": {
        "group": "vscode"
      }
    }
  ]
}

```

Reply:
I've tried that on main as of a few minutes ago, and it seems to work perfectly, without either the `configurationProperties.dependsOn?.length` check (removed by #180546), or the `liveDependencies.delete` which was removed in 3b80fb233 as part of this pull request.

But I've also tried checking out 3b80fb233^, and getting rid of the `configurationProperties.dependsOn?.length`, and then I *am* able to repro - and that makes no sense to me. The `liveDependencies.delete` doesn't affect the result; it's broken without the `configurationProperties.dependsOn?.length` change, and fixed with it, whether or not we do the `liveDependencies.delete` (which makes sense to me, because liveDependencies is 100% dead at the point you added the delete; nothing will ever read it again).

But I'm going to keep debugging it, because I don't understand either why there was a problem then, or why there isn't a problem now...

Reply:
Yes I agree now I understand why `liveDependencies.delete` is not needed. 

When I tested your most recently merged PR in oss, I wasn't seeing the bug where deleting a dependency task and re running the overarching one was not restarting the terminated child. 

I think that was needed because before your change (which checks if a task is running regardless of whether or not `instancePolicy` is set), we were just disallowing that from getting run again 🤷🏼‍♀️

Reply:
Ok - I see what happened. 

Tasks satisfying `_isEmptyTask` (such as `Start demo` in your example) never get added to `_instances`. But prior to #180546 a task without an entry in `_instances` would end up with `validInstance` false. That seems wrong, but prior to #180617 it was ok, because such tasks would never get an entry in `_activeTasks` either, and without an `_activeTasks` entry,  `validInstance` was completely ignored.

Then #180617 changed it so that such tasks *do* get an entry in _activeTasks (from the time that _executeTask is called, until all of their dependencies finish), so `terminalData` is non null, and `validInstance` is false, and we would prevent the task from running a second time. So that's why the bug appeared. At this point the `configurationProperties.dependsOn?.length` check did work around that issue, but also completely disabled the `instanceLimit` checks for any task with dependencies.

But then #180546 fixed it so that if there's no entry in `_instances`, `validInstance` is true (if you have no running instances, you can't possibly have hit the `instanceLimit`), and everything was fine again, without needing the `configurationProperties.dependsOn?.length` check.

So I think everything is ok now that both pull requests have been merged.

Reply:
Thanks for investigating and yeah that all makes sense to me. agreed - based on my testing everything seems to be in a good state. Thank you for the fixes - this is not an easy code area to understand 😅 .